### ISSUE TITLE
rul_hybrid generation crashes with a C-extension refcount error on Linux + torch 2.8

### ISSUE THREAD

**Issue by FelixSaretzky** [FelixSaretzky/xLSTM-RUL#7]
## What happens

`python -m rulbench.synthetic.rul_hybrid --out data/hybrid_train_w16.h5 --n 500 --seed 116 --n-sensors 16` dies reproducibly on a fresh cloud box with

```
Fatal Python error: none_dealloc: deallocating None: bug likely caused by a refcount error in a C extension
```

at ~unit 230 of 500 — two runs crashed with byte-identical partial output (352632 bytes), so the crash point is deterministic.

Environment: Ubuntu 22.04, Python 3.11.13, torch 2.8.0+cu128 (pytorch/pytorch:2.8.0-cuda12.8-cudnn9-devel), dotime 0.1.3, numpy 2.x. torch is in the process via dotime's samplers; rul_hybrid itself is numpy/h5py only.

## What narrows it down

- w3 and w8 at n=500 complete fine; only the widest arm (16 sensors, most sampler calls per unit) hits the crash mid-run.
- n=250 at width 16 completes, and the same abort then fires at interpreter shutdown ("runtime state: finalizing, Garbage-collecting"), after the H5 is written and closed. The trigger is therefore a GC cycle over a corrupted refcount that accumulates with allocation volume — mid-run GC kills big runs, exit GC is harmless.
- Not reproducible locally (macOS arm64, Python 3.12): 1000-unit generation ran clean on 2026-08-06.

## Workaround used for the 2026-08-18 pretraining run

Width 16 generated as two 250-unit files (seeds 1160/1161) in separate processes, passed to training with `--train-weights 0.5 0.5` — same total sampling mass for the arm. Recorded in the run's run.json.

## Open

- Minimal repro to decide where this belongs upstream (dotime's torch sampling path vs torch itself); check whether Python 3.12 or a different torch build on Linux avoids it.
- Hardening: a crashed generation leaves a partial .h5 behind and the "skip if file exists" logic in the smoke script would silently reuse it. Write to a temp name and rename on success.

**Comment by FelixSaretzky** [comment]
Root cause found and proven on the affected box (48020475, Ubuntu 22.04):

| environment | torch | numpy | result |
|---|---|---|---|
| Python 3.11.13 (pytorch docker image) | 2.8.0+cu128 | 2.3.2 | crash at ~unit 230, twice, byte-identical |
| Python 3.11.13, fresh venv | 2.13.0+cpu | 2.4.6 | crash |
| **Python 3.12.13, conda env** | 2.13.0+cpu | 2.5.2 | **completes, exit 0** |
| Python 3.12.2 (macOS, dev machine) | 2.13.0 | 2.5.1 | never crashed, any scale |

Same command, same seed everywhere: `rul_hybrid --n 500 --seed 116 --n-sensors 16`.

The crash is an over-decref of `None` in some C extension on the sampling path. Since CPython 3.12, `None` is immortal (PEP 683), so the extra decref is a no-op there; on 3.11 the refcount reaches zero and the next GC pass aborts the interpreter — mid-run for large generations, at shutdown (harmless, after the file is closed) for small ones. torch and numpy versions are exonerated by rows 2-3.

Consequences:
- Generation on cloud boxes must run on Python >= 3.12. The `pytorch/pytorch` docker images ship 3.11, so either a conda env on the box (`conda create -n py312 python=3.12`) or a different base image. Whether to raise `requires-python` to 3.12 for the package is open for discussion.
- The temp-file-plus-rename hardening for `write_hybrid_dataset` prevents partial files regardless.
- Which extension actually over-decrefs remains unattributed (would need a CPython debug build); not blocking, since 3.12 neutralises it by design.

**Comment by FelixSaretzky** [comment]
Resolved. Root cause is the None over-decref described above: fatal on Python 3.11, neutralised by immortal objects on Python 3.12+ (proof matrix in the previous comment). What landed:

- 85de2fb writes hybrid datasets to a temp file and renames on success, so a crashed generation can never leave a partial HDF5 that skip-if-exists would reuse.
- Operational rule: run generation on Python >= 3.12 on cloud boxes — the pytorch docker images ship 3.11, so create a conda env there (`conda create -n py312 python=3.12`). Verified end to end: the 250-unit chunk regeneration on the py312 env completed with exit 0.

Attributing the over-decref to the specific extension would need a CPython debug build and buys nothing while 3.12 neutralises it by design — not pursued.
